PLANNER-2323 Constraint Streams: map() and distinct() (#1230)
kiegroup/optaplanner: optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

|4          |  Quad|`QuadConstraintStream<A, B, C, D>`
|===

[[constraintStreamsHigherCardinalities]]
=== Achieving higher cardinalities

OptaPlanner currently does not support constraint stream cardinalities higher than 4.

> However, with <<constraintStreamsMappingTuples,tuple mapping>> it is easy to achieve effectively infinite cardinality:

Reviewer: docs and blogs should avoid "it is easy" (although I suspect more docs chapter I wrote suffer from this).

"it is easy" is subjective - it depends on previous experience and understanding of context.
Worse, it also implies an intelligence test for the reader. -> If you can't understand this quickly, you're an idiot. => OptaPlanner is too difficult for me, I need to look for a simpler solution.

[source,java,options="nowrap"]
----
    private Constraint pentaStreamExample(ConstraintFactory factory) {
        return factory.from(Shift.class) // UniConstraintStream<Shift>
                .join(Shift.class)       // BiConstraintStream<Shift, Shift>
                .join(Shift.class)       // TriConstraintStream<Shift, Shift, Shift>
                .join(Shift.class)       // QuadConstraintStream<Shift, Shift, Shift, Shift>
                .map(Tuple::of)          // UniConstraintStream<Tuple<Shift, Shift, Shift, Shift>>

>                 .join(Shift.class)       // BiConstraintStream<Tuple<Shift, Shift, Shift, Shift>, Shift>

Reviewer: the class Tuple doesn't exist, that makes it a bit clumsy to copy-paste. commons-lang3 has Triple and Pair, but not for 4...

                ...                      // This BiConstraintStream carries 5 Shift elements.
    }
----

[NOTE]
====
OptaPlanner does not provide any `Tuple` implementations out of the box.

> We recommend using one of the myriad freely available 3rd party implementations.

Reviewer: Avoid "we" in docs. Avoid the tense of "using" in favor of active:

```suggestion
It's recommended to use one of the myriad freely available 3rd party implementations.
```

Ps: What does "myriad" mean? Many docs readers aren't native English speakers. It's ok to use difficult words if they describe an property of the behavior (such as "bijective"), as it is explained immediately. But in normal sentences, like this, maybe we should avoid uncommon English words without explanation unless they bring enough value. Wdyt?

the stream.
====

[[constraintStreamsMappingTuples]]
=== Mapping tuples

Mapping enables you to transform each tuple in a constraint stream by applying a mapping function to it.
The result of such mapping is `UniConstraintStream` of the mapped tuples.

[source,java,options="nowrap"]
----
    private Constraint computerWithBiggestProcess(ConstraintFactory constraintFactory) {
        return constraintFactory.from(CloudProcess.class) // UniConstraintStream<CloudProcess>
                .map(CloudProcess::getComputer)           // UniConstraintStream<CloudComputer>
                ...
    }
----

[NOTE]
====
In the example above, the mapping function produces duplicate tuples if two different `CloudProcess`es share a single `CloudComputer`.
That is, such `CloudComputer` will appear in the resulting constraint stream twice.
See <<constraintStreamsDistinct,`distinct()`>> for how to deal with duplicate tuples.
====

[[constraintStreamsMappingFunction]]

> ==== Designing the mapping function

Reviewer: (soft) the id differs from the title - so far the id, was usually <namespace of out section if needed> + <camelCase of title or at least the first few words of it>
I'd keep the title and the id very similar, so it's clear they are the same section.

> When designing the mapping function, we recommend you follow these guidelines for optimal performance:

Reviewer: ```suggestion
When designing the mapping function, follow these guidelines for optimal performance:
```

* Keep the function pure.
  The mapping function should only depend on its input.
  That is, given the same input, it always returns the same output.
* Keep the function simple.
  The function should only take the argument and return another simple value,
  without performing any significant computation.
  Not following this recommendation may result in performance loss,
  as the function may be called frequently.
* Keep the function bijective.
  No two input tuples should map to the same output tuple, or to tuples that are equal.

>   Not following this recommendation will create a constraint stream with duplicate tuples,

Reviewer: ```suggestion
  Not following this recommendation creates a constraint stream with duplicate tuples,
```

  and may force you to use <<constraintStreamsDistinct,`distinct()`>> later.
* Use immutable data carriers.
  The tuples returned by the mapping function should be immutable and identified by their contents and nothing else.
  If two tuples carry objects which equal one another,
  those two tuples should likewise equal and preferably be the same instance.

>   https://openjdk.java.net/jeps/395[Java's Records] are an ideal tool for the purpose.

Reviewer: Same comment here as above -> only if we tested it and we are sure it's a good recommendation. Otherwise, let's leave it out for now.

[[constraintStreamsDistinct]]

> ==== Dealing with duplicate tuples

Reviewer: I'd love to see the word "distinct" in this section header as that's the operation it's talking about.
In case I use it (or I inherit a piece of code that uses distinct()) and I want to fully understand what it does by reading the docs, so I need to find it in the TOC.

As a general rule, tuples in constraint streams will be distinct.
That is, no two tuples that equal one another.
However, certain operations such as <<constraintStreamsMappingTuples,tuple mapping>> may produce constraint streams
where that is not true.

If a constraint stream produces duplicate tuples, you can use the `distinct()` building block
to have the duplicate copies eliminated.

[source,java,options="nowrap"]
----
    private Constraint computerWithBiggestProcess(ConstraintFactory constraintFactory) {
        return constraintFactory.from(CloudProcess.class) // UniConstraintStream<CloudProcess>
                .map(CloudProcess::getComputer)           // UniConstraintStream<CloudComputer>
                .distinct()                               // The same, each CloudComputer just once.
                ...
    }
----

[NOTE]
====
There is a performance cost to `distinct()`.

> For optimal performance, make sure that your constraint streams never produce duplicate tuples.

Reviewer: ```suggestion
For optimal performance, don't use constraint stream operations that produce duplicate tuples, to avoid the need to call `distinct()`.
```
